src/host/inputBuffer.cpp

```diff
@@ -261,47 +402,105 @@ void InputBuffer::PassThroughWin32MouseRequest(bool enable)
     {
-        if (_storage.empty())
+        auto event = IInputEvent::Create((*it)->ToInputRecord());
+
+        if (event->EventType() == InputEventType::KeyEvent)
         {
-            if (!WaitForData)
+            const auto keyEvent = static_cast<KeyEvent*>(event.get());
+            WORD repeat = 1;
+
+            // for stream reads we need to split any key events that have been coalesced
+            if (Stream)
             {
-                return STATUS_SUCCESS;
+                repeat = keyEvent->GetRepeatCount();
+                keyEvent->SetRepeatCount(1);
             }
-            return CONSOLE_STATUS_WAIT;
-        }
 
-        // read from buffer
-        std::deque<std::unique_ptr<IInputEvent>> events;
-        size_t eventsRead;
-        bool resetWaitEvent;
-        _ReadBuffer(events,
-                    AmountToRead,
-                    eventsRead,
-                    Peek,
-                    resetWaitEvent,
-                    Unicode,
-                    Stream);
-
-        // copy events to outEvents
-        while (!events.empty())
-        {
-            OutEvents.push_back(std::move(events.front()));
-            events.pop_front();
-        }
+            if (Unicode)
+            {
+                do
+                {
+                    OutEvents.push_back(std::make_unique<KeyEvent>(*keyEvent));
+                    repeat--;
+                } while (repeat > 0 && OutEvents.size() < AmountToRead);
+            }
+            else
+            {
+                const auto wch = keyEvent->GetCharData();
+
+                char buffer[8];
+                const auto length = WideCharToMultiByte(cp, 0, &wch, 1, &buffer[0], sizeof(buffer), nullptr, nullptr);
+                THROW_LAST_ERROR_IF(length <= 0);
+
+                const std::string_view str{ &buffer[0], gsl::narrow_cast<size_t>(length) };
+
+                do
+                {
+                    for (const auto& ch : str)
+                    {
+                        auto tempEvent = std::make_unique<KeyEvent>(*keyEvent);
+                        tempEvent->SetCharData(ch);
+                        OutEvents.push_back(std::move(tempEvent));
+                    }
+                    repeat--;
+                } while (repeat > 0 && OutEvents.size() < AmountToRead);
+            }
 
-        if (resetWaitEvent)
+            if (repeat && !Peek)
+            {
+                const auto originalKeyEvent = static_cast<KeyEvent*>((*it).get());
+                originalKeyEvent->SetRepeatCount(repeat);
+                break;
+            }
+        }
+        else
         {
-            ServiceLocator::LocateGlobals().hInputEvent.ResetEvent();
+            OutEvents.push_back(std::move(event));
         }
-        return STATUS_SUCCESS;
+
+        ++it;
     }
-    catch (...)
+
+    if (!Peek)
+    {
+        _storage.erase(_storage.begin(), it);
+    }
+
+    Cache(Unicode, OutEvents, AmountToRead);

```

Comment:
if cache is only ever used internally, move it into private?

Reply:
Originally I needed to call these functions externally as well and it's entirely possible I need to do it again in the future. I don't think them being public is hurting the code.